File: cropmandi-ai/frontend/src/types/disease.ts
export type AnalysisStatus =
  | 'success'
  | 'unclear_image'
  | 'crop_mismatch'
  | 'non_plant_image'
  | 'insufficient_evidence'
  | 'service_error'
  | 'plantnet_authentication_error'
  | 'plantnet_rate_limit_error'
  | 'plantnet_timeout'
  | 'plantnet_invalid_response'
  | 'plantnet_unavailable';

export type HealthStatus =
  | 'healthy'
  | 'disease'
  | 'fungal_disease'
  | 'bacterial_disease'
  | 'viral_disease'
  | 'pest_damage'
  | 'nutrient_deficiency'
  | 'physical_damage'
  | 'abiotic_stress'
  | 'mixed_or_multiple'
  | 'unclear'
  | 'non_plant_image'
  | 'insufficient_evidence'
  | 'diseased'
  | 'requires_second_stage'
  | 'not_started'
  | 'not_available';

export type IdentificationStatus = 'identified' | 'probable' | 'low_confidence' | 'unidentified' | 'unavailable';

export type CropMatchStatus = 'match' | 'mismatch' | 'uncertain' | 'not_provided' | 'not_specified';

export interface PlantNetSpeciesCandidate {
  scientific_name: string;
  common_names: string[];
  family?: string | null;
  score: number;
  rank: number;
}

export interface ImageQualityAssessment {
  status: 'acceptable' | 'partially_usable' | 'poor' | 'non_plant_image' | 'insufficient_evidence' | 'unclear';
  score?: number | null;
  original_confidence?: number | null;
  issues: string[];
}

export interface LeafMarginAnalysis {
  type: string;
  original_confidence?: number | null;
  evidence: string;
  reliability: 'usable' | 'unreliable' | 'unavailable' | string;
}

export interface FeatureAnalysis {
  leaf_margin?: LeafMarginAnalysis;
  leaf_shape?: string;
  leaf_apex?: string;
  leaf_base?: string;
  leaf_venation?: string;
  leaf_texture?: string;
  leaf_arrangement?: string;
  stem_features?: string;
  fruit_features?: string;
  flower_features?: string;
  root_features?: string;
}

export interface CropCandidate {
  name: string;
  category: string;
  crop_status?: string;
  gemini_original_probability?: number | null;
  classifier_probability?: number | null;
  combined_probability?: number | null;
  supporting_evidence?: string[];
  contradicting_evidence?: string[];
}

export interface BestCrop {
  name: string;
  category: string;
  crop_status?: string;
  gemini_original_probability?: number | null;
  classifier_probability?: number | null;
  combined_probability?: number | null;
  final_selection_source?: string;
}

export interface AmbiguityInfo {
  status: 'low' | 'moderate' | 'high' | string;
  top_candidate_gap?: number | null;
  message?: string;
}

export interface NextImageRequest {
  needed: boolean;
  suggested_images?: string[];
}

export interface CropRecognition {
  identification_status: IdentificationStatus;
  best_crop: BestCrop;
  ranked_candidates: CropCandidate[];
  feature_analysis?: FeatureAnalysis;
  ambiguity?: AmbiguityInfo;

  // Backward compatibility
  crop_name?: string;
  confidence?: number | null;
}

export interface CropComparison {
  user_selected_crop?: string | null;
  detected_best_crop: string;
  match_status: CropMatchStatus;
  reason: string;
}

export interface PlantPartInfo {
  name: string;
  confidence?: number | null;
}

export interface HealthAssessment {
  status: HealthStatus | string;
  confidence?: number | null;
  visible_evidence: string[];
}

export interface PrimaryDiagnosis {
  name: string;
  category?: string;
  confidence?: number | null;
  evidence: string[];
}

export interface AlternativeDiagnosis {
  name: string;
  category?: string;
  confidence?: number | null;
  reason?: string;
  distinguishing_evidence?: string;
}

export interface ChemicalControlGuidance {
  provided: boolean;
  message: string;
}

export interface ModelInfo {
  provider: string;
  model_name: string;
  project?: string;
  prompt_version?: string;
  temperature?: number;
  request_timestamp?: string;
}

export interface ValidationWarning {
  field: string;
  issue: string;
  action: string;
}

export interface CropInfo {
  name: string;
  confidence?: number | null;
}

export interface DiseaseInfo {
  name: string;
  confidence?: number | null;
}

export interface DiseaseAnalysisResult {
  analysis_status?: AnalysisStatus | string;
  provider?: string;
  selected_crop?: string | null;
  detected_crop?: string | null;
  detected_scientific_name?: string | null;
  crop_category?: string | null;
  plantnet_score?: number | null;
  identification_status?: IdentificationStatus | string;
  plantnet_results?: PlantNetSpeciesCandidate[];
  crop_match_status?: CropMatchStatus | string;
  disease_status?: string;

  crop?: CropInfo | string;
  plant_part?: PlantPartInfo | string;
  health_status?: HealthStatus | string;
  disease?: DiseaseInfo;
  symptoms?: string[];
  possible_causes?: string[];
  management?: string[];
  prevention?: string[];
  risk_level?: string;
  limitations?: string[];
  disclaimer?: string;

  // Optional and backward compatibility fields
  image_quality?: ImageQualityAssessment;
  plant_detected?: boolean;
  crop_recognition?: CropRecognition;
  crop_comparison?: CropComparison;
  next_image_request?: NextImageRequest;
  health_assessment?: HealthAssessment;
  primary_diagnosis?: PrimaryDiagnosis;
  alternative_diagnoses?: AlternativeDiagnosis[];
  immediate_actions?: string[];
  chemical_control_guidance?: ChemicalControlGuidance;
  model_disclaimer?: string;
  language?: string;
  validation_warnings?: ValidationWarning[];
  selected_crop_comparison?: CropComparison;
  alternative_possibilities?: AlternativeDiagnosis[];
}

export interface ImageMetadata {
  original_filename: string;
  mime_type: string;
  size_bytes: number;
  sha256: string;
  storage_reference: string;
}

export interface DiseaseAnalysisResponse {
  message: string;
  analysis_id: string;
  created_at: string;
  provider?: string;
  model?: ModelInfo;
  result: DiseaseAnalysisResult;
  warnings?: string[];
  disclaimer?: string;
}

export interface DiseaseHistoryItem {
  analysis_id: string;
  created_at: string;
  provider?: string;
  selected_crop?: string | null;
  detected_crop: string;
  detected_scientific_name?: string | null;
  detected_crop_category?: string;
  plantnet_score?: number | null;
  crop_status?: string;
  identification_status?: string;
  crop_match_status?: string;
  plantnet_results?: PlantNetSpeciesCandidate[];
  plant_part: string;
  health_status: string;
  disease_status?: string;
  ranked_candidates?: CropCandidate[];
  original_confidence: {
    crop?: number | null;
    plantnet_score?: number | null;
    plant_part?: number | null;
    health_status?: number | null;
    primary_diagnosis?: number | null;
  };
  gemini_original_probability?: number | null;
  classifier_probability?: number | null;
  combined_probability?: number | null;
  feature_analysis?: FeatureAnalysis;
  ambiguity?: AmbiguityInfo;
  next_image_request?: NextImageRequest;
  primary_diagnosis: {
    name: string;
    category?: string;
    confidence?: number | null;
    evidence: string[];
  };
  symptoms: string[];
  possible_causes: string[];
  immediate_actions: string[];
  prevention: string[];
  image_quality?: Record<string, any>;
  has_image?: boolean;
  image_url?: string;
  image_metadata?: ImageMetadata;
  model?: ModelInfo;
  language?: string;
  warnings?: string[];
  disclaimer?: string;

  // Compatibility aliases
  crop?: string;
  user_id?: string;
  user_symptoms?: string | null;
  user_notes?: string | null;
  image?: ImageMetadata;
  result?: DiseaseAnalysisResult;
}

export interface DiseaseHistoryListResponse {
  analyses: DiseaseHistoryItem[];
  total_count: number;
  user_id: string;
}

File: cropmandi-ai/backend/app/schemas/disease.py
from typing import List, Optional, Dict, Any, Literal, Union
from pydantic import BaseModel, Field
from datetime import datetime

# Enums / Literal Types
AnalysisStatus = Literal[
    "success",
    "uncertain",
    "unclear_image",
    "crop_mismatch",
    "non_plant_image",
    "insufficient_evidence",
    "service_error",
    "plantnet_authentication_error",
    "plantnet_rate_limit_error",
    "plantnet_timeout",
    "plantnet_invalid_response",
    "plantnet_unavailable"
]

HealthStatus = Literal[
    "healthy",
    "diseased",
    "disease",
    "uncertain",
    "fungal_disease",
    "bacterial_disease",
    "viral_disease",
    "pest_damage",
    "nutrient_deficiency",
    "physical_damage",
    "abiotic_stress",
    "mixed_or_multiple",
    "unclear",
    "non_plant_image",
    "insufficient_evidence",
    "requires_second_stage",
    "not_started",
    "not_available"
]

IdentificationStatus = Literal[
    "identified",
    "probable",
    "low_confidence",
    "unidentified",
    "unavailable"
]

CropMatchStatus = Literal[
    "match",
    "mismatch",
    "uncertain",
    "not_provided",
    "not_specified"
]

CropCategory = Literal[
    "vegetable",
    "fruit",
    "cereal",
    "millet",
    "pulse",
    "oilseed",
    "spice",
    "plantation",
    "fiber",
    "tuber",
    "root",
    "leafy",
    "fodder",
    "cultivated_plant",
    "wild_plant",
    "ornamental",
    "unknown"
]


class PlantNetSpeciesCandidate(BaseModel):
    scientific_name: str = Field(description="Scientific botanical name e.g. Solanum lycopersicum L.")
    common_names: List[str] = Field(default_factory=list, description="Common names e.g. ['Tomato', 'Garden tomato']")
    family: Optional[str] = Field(default=None, description="Botanical family e.g. Solanaceae")
    score: float = Field(default=0.0, ge=0.0, description="PlantNet identification score")
    rank: int = Field(default=1, description="Rank in results list")


class ImageQualityAssessment(BaseModel):
    status: str = Field(default="acceptable", description="acceptable, partially_usable, poor, non_plant_image, insufficient_evidence")
    original_confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    score: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    issues: List[str] = Field(default_factory=list)


class LeafMarginAnalysis(BaseModel):
    type: str = Field(default="unavailable", description="serrated, dentate, crenate, smooth/entire, lobed, undulate, irregular_damaged, unavailable")
    original_confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    evidence: str = Field(default="")
    reliability: str = Field(default="usable", description="usable, unreliable, unavailable")


class FeatureAnalysis(BaseModel):
    leaf_margin: LeafMarginAnalysis = Field(default_factory=LeafMarginAnalysis)
    leaf_shape: str = Field(default="")
    leaf_apex: str = Field(default="")
    leaf_base: str = Field(default="")
    leaf_venation: str = Field(default="")
    leaf_texture: str = Field(default="")
    leaf_arrangement: str = Field(default="")
    stem_features: str = Field(default="")
    fruit_features: str = Field(default="")
    flower_features: str = Field(default="")
    root_features: str = Field(default="")


class CropCandidate(BaseModel):
    name: str = Field(description="Crop name e.g. Tomato, Groundnut, Maize, Mango")
    category: str = Field(default="vegetable", description="vegetable, fruit, cereal, millet, pulse, oilseed, spice, plantation...")
    crop_status: str = Field(default="recognized", description="recognized or recognized_outside_configured_vocabulary")
    gemini_original_probability: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    classifier_probability: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    combined_probability: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    supporting_evidence: List[str] = Field(default_factory=list)
    contradicting_evidence: List[str] = Field(default_factory=list)


class BestCrop(BaseModel):
    name: str = Field(description="Selected highest-probability crop name")
    category: str = Field(default="vegetable")
    crop_status: str = Field(default="recognized", description="recognized or recognized_outside_configured_vocabulary")
    gemini_original_probability: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    classifier_probability: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    combined_probability: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    final_selection_source: str = Field(default="plantnet_identification_score")


class AmbiguityInfo(BaseModel):
    status: str = Field(default="low", description="low, moderate, high")
    top_candidate_gap: Optional[float] = Field(default=None)
    message: str = Field(default="")


class CropRecognition(BaseModel):
    identification_status: IdentificationStatus = Field(default="probable")
    best_crop: BestCrop
    ranked_candidates: List[CropCandidate] = Field(default_factory=list)
    feature_analysis: FeatureAnalysis = Field(default_factory=FeatureAnalysis)
    ambiguity: AmbiguityInfo = Field(default_factory=AmbiguityInfo)

    @property
    def crop_name(self) -> str:
        return self.best_crop.name

    @property
    def confidence(self) -> Optional[float]:
        return self.best_crop.combined_probability or self.best_crop.gemini_original_probability


class CropComparison(BaseModel):
    user_selected_crop: Optional[str] = None
    detected_best_crop: str
    match_status: CropMatchStatus = Field(default="not_provided")
    reason: str = Field(default="")


class NextImageRequest(BaseModel):
    needed: bool = Field(default=False)
    suggested_images: List[str] = Field(default_factory=list)


class PlantPartInfo(BaseModel):
    name: str = Field(default="Leaf", description="Leaf, Stem, Fruit, Root, Flower, Whole plant, Multiple parts, Unknown")
    confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)


class HealthAssessment(BaseModel):
    status: str = Field(default="requires_second_stage")
    confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    visible_evidence: List[str] = Field(default_factory=list)


class PrimaryDiagnosis(BaseModel):
    name: str = Field(default="Plant Species Identified (Expert Diagnosis Pending)", description="Condition name")
    category: str = Field(default="requires_second_stage")
    confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    evidence: List[str] = Field(default_factory=list)


class AlternativeDiagnosis(BaseModel):
    name: str
    category: Optional[str] = None
    confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    reason: Optional[str] = None
    distinguishing_evidence: Optional[str] = None


class ChemicalControlGuidance(BaseModel):
    provided: bool = Field(default=False)
    message: str = Field(
        default="Consult a local agricultural extension officer or Krishi Vigyan Kendra (KVK) for approved regional products."
    )


class ValidationWarning(BaseModel):
    field: str
    issue: str
    action: str


class ModelInfo(BaseModel):
    provider: str = Field(default="PlantNet")
    model_name: str = Field(default="PlantNet-v2")
    project: str = Field(default="all")
    prompt_version: Optional[str] = Field(default=None)
    request_timestamp: Optional[str] = Field(default=None)


class CropInfo(BaseModel):
    name: str = Field(default="Unidentified Plant", description="Detected crop name e.g. Tomato, Groundnut")
    confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)


class DiseaseInfo(BaseModel):
    name: str = Field(default="Expert diagnosis pending", description="Diagnosed disease")
    confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)


class DiseaseAnalysisResult(BaseModel):
    analysis_status: str = Field(default="success", description="success | insufficient_evidence | non_plant_image | service_error")
    provider: str = Field(default="PlantNet")
    selected_crop: Optional[str] = None
    detected_crop: Optional[str] = None
    detected_scientific_name: Optional[str] = None
    crop_category: Optional[str] = None
    plantnet_score: Optional[float] = Field(default=None, description="PlantNet identification score")
    identification_status: str = Field(default="identified", description="identified | probable | low_confidence | unavailable")
    plantnet_results: List[PlantNetSpeciesCandidate] = Field(default_factory=list)
    crop_match_status: str = Field(default="not_provided", description="match | mismatch | uncertain | not_provided")
    disease_status: str = Field(default="requires_second_stage", description="not_started | unclear | requires_second_stage")

    crop: CropInfo = Field(default_factory=CropInfo)
    plant_part: str = Field(default="Leaf")
    health_status: str = Field(default="requires_second_stage")
    disease: DiseaseInfo = Field(default_factory=DiseaseInfo)
    symptoms: List[str] = Field(default_factory=list)
    possible_causes: List[str] = Field(default_factory=list)
    management: List[str] = Field(default_factory=list)
    prevention: List[str] = Field(default_factory=list)
    risk_level: str = Field(default="low", description="low | medium | high | uncertain")
    limitations: List[str] = Field(default_factory=list)
    disclaimer: str = Field(
        default="PlantNet identification is an AI-assisted preliminary species identification and is not a guaranteed disease diagnosis."
    )

    # Backward compatibility fields
    plant_detected: bool = Field(default=True)
    image_quality: Optional[ImageQualityAssessment] = Field(default_factory=ImageQualityAssessment)
    crop_recognition: Optional[CropRecognition] = None
    crop_comparison: Optional[CropComparison] = None
    next_image_request: Optional[NextImageRequest] = Field(default_factory=NextImageRequest)
    health_assessment: Optional[HealthAssessment] = None
    primary_diagnosis: Optional[PrimaryDiagnosis] = None
    alternative_diagnoses: List[AlternativeDiagnosis] = Field(default_factory=list)
    immediate_actions: List[str] = Field(default_factory=list)
    chemical_control_guidance: Optional[ChemicalControlGuidance] = Field(default_factory=ChemicalControlGuidance)
    model_disclaimer: str = Field(
        default="Plant identified by PlantNet. Disease diagnosis requires a separate disease-analysis model or expert confirmation."
    )
    language: str = Field(default="en")
    validation_warnings: List[ValidationWarning] = Field(default_factory=list)


class ImageMetadata(BaseModel):
    original_filename: str
    mime_type: str
    size_bytes: int
    sha256: str
    storage_reference: str


class DiseaseAnalysisResponse(BaseModel):
    message: str = Field(default="Crop image identified successfully via PlantNet.")
    analysis_id: str
    created_at: str
    provider: str = Field(default="PlantNet")
    model: Optional[ModelInfo] = None
    result: DiseaseAnalysisResult
    warnings: List[str] = Field(default_factory=list)
    disclaimer: str = Field(
        default="PlantNet identification is an AI-assisted preliminary identification and is not a guaranteed disease diagnosis."
    )


class DiseaseHistoryItem(BaseModel):
    analysis_id: str
    created_at: str
    provider: str = "PlantNet"
    selected_crop: Optional[str] = None
    detected_crop: Optional[str] = None
    detected_scientific_name: Optional[str] = None
    detected_crop_category: str = "vegetable"
    plantnet_score: Optional[float] = None
    crop_status: str = "recognized"
    identification_status: str = "identified"
    crop_match_status: str = "not_provided"
    plantnet_results: List[Dict[str, Any]] = Field(default_factory=list)
    plant_part: str = "Leaf"
    health_status: str = "requires_second_stage"
    disease_status: str = "requires_second_stage"
    ranked_candidates: List[Dict[str, Any]] = Field(default_factory=list)
    original_confidence: Dict[str, Optional[float]] = Field(default_factory=dict)
    primary_diagnosis: Dict[str, Any] = Field(default_factory=dict)
    feature_analysis: Optional[Dict[str, Any]] = None
    ambiguity: Optional[Dict[str, Any]] = None
    next_image_request: Optional[Dict[str, Any]] = None
    symptoms: List[str] = Field(default_factory=list)
    possible_causes: List[str] = Field(default_factory=list)
    immediate_actions: List[str] = Field(default_factory=list)
    prevention: List[str] = Field(default_factory=list)
    image_quality: Dict[str, Any] = Field(default_factory=dict)
    has_image: bool = True
    image_url: Optional[str] = None
    image_metadata: Optional[ImageMetadata] = None
    model: Optional[ModelInfo] = None
    language: str = "en"
    warnings: List[str] = Field(default_factory=list)
    disclaimer: str = "PlantNet identification is an AI-assisted preliminary identification and is not a guaranteed disease diagnosis."

    # Compatibility aliases
    crop: Optional[str] = None
    user_id: Optional[str] = None
    user_symptoms: Optional[str] = None
    user_notes: Optional[str] = None
    image: Optional[ImageMetadata] = None
    result: Optional[DiseaseAnalysisResult] = None


class DiseaseHistoryListResponse(BaseModel):
    analyses: List[DiseaseHistoryItem]
    total_count: int
    user_id: str
